Common/Data/Consolidators/RangeConsolidator.cs

```diff
@@ -0,0 +1,223 @@
+        /// <param name="volume">Volume of the given data</param>
+        protected override void UpdateBar(DateTime time, decimal currentValue, decimal volume)
+        {
+            if (currentValue > CurrentRangeBar.Low + 2 * RangeSize)
+            {
+                Rising(time, currentValue, volume);
+            }
+            else if (currentValue < CurrentRangeBar.High - (2 * RangeSize))
+            {
+                Falling(time, currentValue, volume);
+            }
+            else
+            {
+                CurrentRangeBar.Update(time, currentValue, volume);
+            }
+        }
+
+        /// <summary>
+        /// Checks if the current RangeBar being created has closed. If that is the case, it consolidates
+        /// the bar, resets the current RangeBar and stores the needed information from the last bar to
+        /// create the next bar
+        /// <see cref="UpdateBar(DateTime, decimal, decimal)"/>
+        /// </summary>
+        protected override void CheckIfBarIsClosed()
+        {
+            if (CurrentRangeBar != null && CurrentRangeBar.IsClosed)
+            {
+                OnDataConsolidated(CurrentRangeBar);
+                CurrentRangeBar = null;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new bar with the given data
+        /// </summary>
+        /// <param name="data">The new data for the bar</param>
+        protected override void CreateNewBar(IBaseData data)
+        {
+            var currentValue = Selector(data);
+            var volume = VolumeSelector(data);
+            var open = currentValue;
+
+            if (_firstTick)
+            {
+                _minimumPriceVariation = SymbolPropertiesDatabase.FromDataFolder().GetSymbolProperties(data.Symbol.ID.Market, data.Symbol, data.Symbol.ID.SecurityType, "USD").MinimumPriceVariation;
+                RangeSize = _minimumPriceVariation * Range;
+                open = Math.Ceiling(open / RangeSize) * RangeSize;
+                _firstTick = false;
+            }
+
+            CurrentRangeBar = new RangeBar(data.Symbol, data.Time, RangeSize, open, volume);
+        }
+
+        /// <summary>
+        /// Creates intermediate/phantom RangeBar's when the price is rising
+        /// </summary>
+        /// <param name="time">Time of the given data</param>
+        /// <param name="currentValue">Value of the given data</param>
+        /// <param name="volume">Volume of the given data</param>
+        private void Rising(DateTime time, decimal currentValue, decimal volume)
+        {
+            while (currentValue - CurrentRangeBar.Low > RangeSize)
+            {
+                CurrentRangeBar.Update(time, currentValue, volume);
+                OnDataConsolidated(CurrentRangeBar);
+                CurrentRangeBar = new RangeBar(CurrentRangeBar.Symbol, CurrentRangeBar.EndTime, RangeSize, CurrentRangeBar.Close + _minimumPriceVariation, 0);
+            }
+            CurrentRangeBar = null;
```

Comment:
Just a thougth but, it could be possible to get by without `Rising` and `Falling`? 
The loop stopping condition might be `while (CurrentBar.IsClosed)` and the open price passed to `new RangeBar()` might be something like `isRising ? CurrentBar.High + _minimumPriceVariation : CurrentBar.Low - _minimumPriceVariation`. 

Then this code might be moved to `UpdateBar` and there would be no need for the 3 cases there, it would be abstracted into just one loop. I think it would handle the 3 cases: (1 and 2) rising avobe/falling bellow `2*range` (more than one bar will be formed) or (3) just one bar will be formed.